server/graphql/common/expenses.ts

```diff
@@ -865,12 +867,56 @@ export async function createExpense(
 
   // Load the payee profile
   const fromCollective = expenseData.fromCollective || (await remoteUser.getCollective());
+
   if (!remoteUser.isAdminOfCollective(fromCollective)) {
     throw new ValidationFailed('You must be an admin of the account to submit an expense in its name');
   } else if (!fromCollective.canBeUsedAsPayoutProfile()) {
     throw new ValidationFailed('This account cannot be used for payouts');
   }
 
+  // Check if expensesType hasn't been disabled by the collective or the host
+  switch (expenseData.type) {
+    case EXPENSE_TYPE.INVOICE:
+        // Check if host allow expense of type invoice
+        if (
+            collective?.settings?.expenseTypes?.hasInvoice === false
+        ) {
+          throw new FeatureNotSupportedForCollective('Expense of type invoice has been disabled by the collective');
+        } else if (
+            !isBoolean(collective.settings?.expenseTypes?.hasInvoice)
+            && collective.host?.settings.expenseTypes.hasInvoice === false
+        ) {
+          throw new FeatureNotSupportedForCollective('Expense of type invoice has been disabled by the host');
+        }
+      break
+    case EXPENSE_TYPE.GRANT:
+      // Check if host allow expense of type invoice
+      if (
+          collective.settings?.expenseTypes?.hasGrant === false
+      ) {
+        throw new FeatureNotSupportedForCollective('Expense of type grant has been disabled by the collective');
+      } else if (
+          !isBoolean(collective.settings?.expenseTypes?.hasGrant)
+          && collective.host.settings?.expenseTypes?.hasGrant === false
+      ) {
+        throw new FeatureNotSupportedForCollective('Expense of type grant has been disabled by the host');
+      }
+      break
+    case EXPENSE_TYPE.RECEIPT:
+      // Check if host has expense of type receipt
+      if (
+          collective.settings?.expenseTypes?.hasReceipt === false
+      ) {
+        throw new FeatureNotSupportedForCollective('Expense of type receipt has been disabled by the collective');
+      } else if (
+          !isBoolean(collective.settings?.expenseTypes?.hasReceipt)
+          && collective.host.settings?.expenseTypes?.hasReceipt === false
+      ) {
+        throw new FeatureNotSupportedForCollective('Expense of type receipt has been disabled by the host');
+      }
+      break
+  }
+
```

Comment:
From what I understand, we only cover the `createExpense` mutation here. As mentioned during the call, we want to cover the `editExpense` one as well **when changing the type of the expense**. In other words, editing an expense that has an invalid type shouldn't block, but changing the type of an expense from `RECEIPT` to `GRANT` when grants are forbidden shouldn't be allowed.

Can we:
1. Move all this code to a `checkExpenseType` helper
2. Re-use it in `editExpense`
3. Bonus: Refactor the code to reduce the duplication